_posts/2021-03-23-activesupport.html

```diff
@@ -0,0 +1,172 @@
+  you can have more than one subscriber listening to <code>process_action.action_controller</code>.
+</p>
+
+{% highlight ruby %}
+# config/initializers/slowlog_subscriber.rb
+
+module ActionController
+  class SlowlogSubscriber < ActiveSupport::Subscriber
+    attach_to :action_controller
+
+    def process_action(message)
+      return if message.duration <= 10
+
+      controller_location = [message.payload[:controller], message.payload[:action]].join("#") # -> ThingsController#something
+      Rails.logger.debug "#{controller_location} was slow (#{message.duration}ms)" 
+    end
+  end
+end
+{% endhighlight %}
+
+<p>
+  You are free to organize this however you want. Decouple publisher/subscriber in different files, chronological or
+  even in different threads.
+</p>
+
+{% highlight ruby %}
+# config/initializers/poor_mans_background_job_subscriber.rb
+
+module ActionController
+  class PMBGJSubscriber < ActiveSupport::Subscriber
+    include Concurrent::Async
+
+    attach_to :action_controller
+
+    def process_action(message)
+      async.background_job(message)
+    end
+
+    def background_job(message)
+      # ...do something expensive with the message in a thread
+      sleep(60)
+    end
+  end
+end
+{% endhighlight %}
+
+<h2>ActiveSupport::Notifications is everywhere today</h2>
+<p>
+  Second, and you probably already guessed this from the example, what makes this most awesome are the ready made
+  messages that are already published <b>today</b>.
+</p>
+<p>
+  Rails for instance uses <code>ActiveSupport::Notifications</code> to publish close to 50(!) different instrumentation events
+  that include data about your application. Data ranging from the HTTP status code of your requests, over which partials
+  were rendered, to more esoteric measurements like the byte range attempted to be read from your ActiveStorage service.
+  Check the <a href="https://guides.rubyonrails.org/active_support_instrumentation.html">instrumentation guide</a>
+  for all the dirty secrets.
+</p>
+
+<h2>ActiveSupport::Notifications is easily extendible</h2>
+<p>
+  Last but not least, you can not only listen to messages others publish, you can publish messages to yourself.
+</p>
+
+{% highlight ruby %}
+ActiveSupport::Notifications.instrument "cool_thing_happened.my_app", {some: :data} do
+  MyApp.do_cool_thing(true)
+end
+{% endhighlight %}
+
+<p>
+  Okay, you're convinced I hope! Now what do people do with <code>ActiveSupport::Notifications</code> out in the world?
+</p>
+
+<h2>Application Health/Performance Monitoring (A*M)</h2>
+<p>
+  Ever wondered how your metrics get to sentry, new relic, datadog or skylight? You guessed it, via <code>ActiveSupport::Notifications</code>.
+</p>
+<p>
+  Now if the main work, publishing tons of messages through a pub-sub architecture, is already done by Rails, surely there are
+  non SaaS (a.k.a. Free Software) options to display this data right? Turns out there aren't really. While evaluating A*M for some
```

Comment:
non-SaaS is a compound adjective. There's also a missing comma before `right`.

```suggestion
  non-SaaS (a.k.a. Free Software) options to display this data, right? Turns out there aren't really. While evaluating A*M for some
```